### ISSUE TITLE
Add to cart redirect won't work if there are notices in WC

### ISSUE THREAD

**Issue by nviitala** [woothemes/woocommerce#37164]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

add_to_cart_action checks that the notice count must be 0 for it to be able to redirect the user to the cart ("0 === wc_notice_count( 'error' )", line 800 in class-wc-form-handler.php).

Lets say we have eg. added the same coupon code twice, before the add to cart
`add_action( 'wp_loaded', function() {
    global $woocommerce;
    $woocommerce->cart->add_discount( 'foo' );
    $woocommerce->cart->add_discount( 'foo' );
});`

Lets say that this current page doesn't output WC notices, so the notices will stay in the cookies.

Now if the user goes to ?add-to-cart=xxx, he won't be redirected to the cart page, even if that option would be enabled in the options.

### Expected behavior

User should be redirected to the cart page even if there are notices.

### Actual behavior

User isn't redirected.

### Steps to reproduce

1. Enable "Redirect to the cart page after successful addition"
2. Add notices to wc, eg trying to add the same coupon twice
`add_action( 'wp_loaded', function() {
    global $woocommerce;
    $woocommerce->cart->add_discount( 'foo' );
    $woocommerce->cart->add_discount( 'foo' );
});`
3. Visit ?add-to-cart=1

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://nyytit.test
Site address (URL): http://nyytit.test
WC Version: 7.4.1
REST API Version: ✔ 7.4.1
WC Blocks Version: ✔ 9.4.4
Action Scheduler Version: ✔ 3.5.4
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: fi
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.46 (Unix) OpenSSL/1.0.2u PHP/7.4.26 mod_wsgi/3.5 Python/2.7.13 mod_fastcgi/mod_fastcgi-SNAP-0910052141 mod_perl/2.0.11 Perl/v5.30.1
PHP Version: 7.4.26
PHP Post Max Size: 265 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.68.0
OpenSSL/1.0.2u

SUHOSIN Installed: –
MySQL Version: 8.0.11
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.4.1
WC Database Prefix: wp_
Total Database Size: 66.24MB
Database Data Size: 41.41MB
Database Index Size: 24.83MB
wp_woocommerce_sessions: Data: 0.38MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.33MB + Index: 0.13MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 6.52MB + Index: 5.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 1.52MB + Index: 0.66MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.41MB + Index: 0.36MB + Engine InnoDB
wp_aioseo_cache: Data: 0.08MB + Index: 0.03MB + Engine InnoDB
wp_aioseo_notifications: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_aioseo_posts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 1.52MB + Index: 1.05MB + Engine InnoDB
wp_gens_raf: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gens_rafmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailchimp_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailchimp_jobs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_options: Data: 2.47MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 7.52MB + Index: 8.03MB + Engine InnoDB
wp_posts: Data: 14.52MB + Index: 0.39MB + Engine InnoDB
wp_rank_math_404_logs: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_rank_math_analytics_gsc: Data: 2.52MB + Index: 4.94MB + Engine InnoDB
wp_rank_math_analytics_inspections: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_rank_math_analytics_objects: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_rank_math_internal_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_rank_math_internal_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_rank_math_redirections: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_rank_math_redirections_cache: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 1.52MB + Index: 1.97MB + Engine InnoDB
wp_users: Data: 0.09MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.08MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.14MB + Index: 0.13MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.31MB + Index: 0.36MB + Engine InnoDB
wp_wc_order_stats: Data: 0.16MB + Index: 0.19MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.11MB + Index: 0.13MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wdr_order_discounts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wdr_order_item_discounts: Data: 0.22MB + Index: 0.19MB + Engine InnoDB
wp_wdr_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpforms_tasks_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpmailsmtp_debug_events: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpmailsmtp_tasks_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

adp-popup: 1
attachment: 145
custom_css: 1
page: 10
post: 8
product: 6
revision: 489
shop_coupon: 5
shop_order: 1185
wp_block: 1
wp_global_styles: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (20) ###

Advanced Popups: by Code Supply Co. – 1.1.3
All-in-One WP Migration URL Extension: by ServMask – 2.60
All-in-One WP Migration: by ServMask – 7.71
Yoast Duplicate Post: by Enrico Battocchi & Team Yoast – 4.5
GTM4WP: by Thomas Geiger – 1.16.2
F4 Shipping Phone and E-Mail for WooCommerce: by FAKTOR VIER – 1.0.17
GenerateBlocks: by Tom Usborne – 1.7.2
Klarna Order Management for WooCommerce: by klarna
krokedil – 1.7.1

Klarna Payments for WooCommerce: by krokedil
klarna
automattic – 3.0.5

Mailchimp for WooCommerce: by Mailchimp – 2.8
OGOShip API for WooCommerce: by OGOShip – 3.7.0
Paytrail for WooCommerce: by Paytrail – 1.3.2
Rank Math SEO: by Rank Math – 1.0.109
SVG Support: by Benbodhi – 2.5.5
WooCommerce Carrier Agents: by Lauri Karisola / markup.fi – 1.6.0
Woo Discount Rules PRO 2.0: by Flycart – 2.5.4
Woo Discount Rules: by Flycart – 2.5.4
WooCommerce: by Automattic – 7.4.1
WP Mail SMTP: by WPForms – 3.7.0
Refer a Friend for WooCommerce PREMIUM: by WPGens – 2.3.18

### Inactive Plugins (9) ###

Advanced Order Export For WooCommerce: by AlgolPlus – 3.3.3
Akismet Anti-Spam: by Automattic – 5.0.2
Block Visibility: by Nick Diego – 2.6.0
Easy SVG Support: by Benjamin Zekavica – 3.3.0
Loco Automatic Translate Addon PRO: by Cool Plugins – 1.4.1
Loco Translate: by Tim Whitlock – 2.6.3
Order Export & Order Import for WooCommerce: by WebToffee – 2.3.0
WooCommerce Conditional Shipping and Payments: by WooCommerce – 1.14.2
WooCommerce Weight Based Shipping: by weightbasedshipping.com – 5.4.1

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
Order datastore: WC_Order_Data_Store_CPT

### WC Pages ###

Shop base: #205 - /shop/
Cart: #206 - /ostoskori/
Checkout: #207 - /kassa/
My account: #208 - /my-account/
Terms and conditions: #209 - /tilaus-ja-sopimusehdot/

### Theme ###

Name: Nyytit Theme
Version: 3.1.0
Author URL: 
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: nyytit/woocommerce/cart/cart-totals.php
nyytit/woocommerce/checkout/form-billing.php
nyytit/woocommerce/checkout/form-checkout.php
nyytit/woocommerce/checkout/order-receipt.php
nyytit/woocommerce/emails/customer-new-account.php

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
new-product-management-experience
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
product-variation-management
settings

Daily Cron: ✔ Next scheduled: 2023-03-11 06:27:53 +00:00
Options: ✔
Notes: 91
Onboarding: completed

### Action Scheduler ###

Complete: 1,529
Oldest: 2023-02-07 16:52:00 +0000
Newest: 2023-03-10 07:23:22 +0000

Failed: 126
Oldest: 2022-09-23 06:03:44 +0000
Newest: 2023-03-06 06:13:17 +0000

Pending: 3
Oldest: 2023-03-11 06:31:04 +0000
Newest: 2023-03-16 07:59:08 +0000

### Status report information ###

Generated at: 2023-03-10 07:35:13 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by barryhughes** [comment]
Thanks for reporting this, @nviitala.

> Lets say that this current page doesn't output WC notices, so the notices will stay in the cookies.

Nice catch and, if I'm understanding correctly, it would seem this could happen even if the current page *does* output notices (such as in your example where the error-generating code runs before the add-to-cart action, ie prior to `wp_loaded:20`).

> User should be redirected to the cart page even if there are notices.

A little less sure about this. It would seem like it might be preferable to render the error notices followed by a prompt such as, *'Proceed to cart anyway?'* Of course, in those cases where the notices are not rendering, that would be problematic.

Can you provide some information about where you are seeing this, to help us chart the best way forward?

**Comment by nviitala** [comment]
@barryhughes Thank you for the prompt check and reply.

>  it would seem this could happen even if the current page does output notices

That is correct

> Can you provide some information about where you are seeing this, to help us chart the best way forward?

We are bringing in users with discount codes in the URL. Sadly, the URL gets modified & user redirected sometimes after the WP loads (eg. removing a double parameter in the url). In this case, the discount code gets applied twice. Same happens if the user would just reload the page.
This can of course be checked with "has_discount" method, but some plugins check it eg on "wp_loaded", when it doesn't necessarily always work (and should be checked at "wp" hook or later instead).
We are redirecting users straight to cart from the landing page via normal links with ?add-to-cart=xxx.

> A little less sure about this. It would seem like it might be preferable to render the error notices followed by a prompt such as, 'Proceed to cart anyway?'

It might be difficult to add a prompt when we are using normal links & buttons with ?add-to-cart url (instead of JS) and custom pages.
The notices are available in the cart page anyway and it is important to notice that the error notices aren't necessarily related to the product itself nor hard errors (eg. couldn't be able to apply the discount code twice).

My suggestion:
Maybe the default behaviour could be improved with a filter? That way the programmer would be able to edit the check to suit the needs of the site.
The default way itself could maybe be to check that if the cart has errors and NO products, then do not redirect to the cart?

**Comment by barryhughes** [comment]
Those are great notes—thank you for sharing. I definitely agree a new filter would be a solid compromise and would be pretty easy (and, more importantly, safe) solution to implement.

**Comment by nviitala** [comment]
@barryhughes Any news regarding this? Updating WC in is difficult because we have to manually fix this every time.
Or would you recommend that I would add the filter for you & make a pull request?

**Comment by barryhughes** [comment]
If you wish to create a pull request, that would be welcome and we'll be happy to review it. Otherwise, this is not currently a high priority from our perspective and I'm sorry to say I can't offer a timeline to resolution.

**Comment by clemrdh** [comment]
Sadly, no updates ?

**Comment by pierorocca** [comment]
Hi @nviitala - thanks for the report.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

This appears to still be relevant, so I've marked it ready for the engineering team to pick up.

What I'm seeing: The classic add-to-cart handler on trunk still blocks redirect behavior whenever any error notice exists, which matches the issue's complaint that notices can prevent redirect-to-cart from working. No related open PR was found in the audit cache.

**Comment by pierorocca** [comment]
Claude Code notes:

Now let me look at the relevant code:


## The Bug

**"Add to cart redirect won't work if there are notices in WC"**

### What's happening

In `class-wc-form-handler.php`, the `add_to_cart_action()` function decides whether to redirect the user to the cart page after a successful add-to-cart. The guard condition is:

```php
if ( 0 === wc_notice_count( 'error' ) ) {
    // redirect to cart
}
```

The problem is that **WooCommerce notices persist in cookies** across page loads. So if *any prior page request* generated an error notice (even completely unrelated to the current add-to-cart action), those stale notices sit in the cookie and block the redirect — even though the product was successfully added.

### Concrete example from the issue

```php
add_action( 'wp_loaded', function() {
    $woocommerce->cart->add_discount( 'foo' );
    $woocommerce->cart->add_discount( 'foo' ); // triggers "coupon already applied" error
});
```

The error notice from the duplicate coupon gets stored in the cookie. Then when the user visits `?add-to-cart=123`, the product is added fine, but the error count is non-zero (from the stale cookie), so the redirect to cart never fires.

`★ Insight ─────────────────────────────────────`
- **WC notices are session/cookie-backed**, not request-scoped. `wc_notice_count()` reads from `WC()->session`, which persists across requests. This is fundamentally different from a per-request flash notice system.
- **The guard condition conflates two orthogonal concerns**: "did *this* add-to-cart action fail?" vs "are there *any* error notices in the session?" The bug is that a stale notice from a previous request poisons the redirect decision.
- **The `wp_loaded` hook** runs before WooCommerce's own add-to-cart processing (which hooks at `wp_loaded:20`), which is why the duplicate-coupon notice appears before the add-to-cart even runs — making it impossible to distinguish "errors from this request" from "errors from before."
`─────────────────────────────────────────────────`

### Proposed fix

The reporter and a WooCommerce core member agreed the cleanest fix is a **filter**, letting developers override the redirect decision. Something like:

```php
$should_redirect = ( 0 === wc_notice_count( 'error' ) );
$should_redirect = apply_filters( 'woocommerce_add_to_cart_redirect_to_cart', $should_redirect );
if ( $should_redirect ) {
    wp_safe_redirect( wc_get_cart_url() );
}
```

### Current status

- Open, labelled as **community contribution** — WooCommerce maintainers have said a PR is welcome but it's not high priority for them to fix internally.
- This is a relatively safe, low-risk change: adding a filter doesn't change default behavior, it just gives developers an escape hatch.

---

**In short**: stale WC notices (stored in cookies/session from previous requests) block the add-to-cart redirect even when the current add-to-cart was successful. The fix is adding a filter to let site developers override the redirect decision.
